sklearn/ensemble/_hist_gradient_boosting/binning.py

```diff
@@ -74,6 +74,10 @@ def _find_binning_thresholds(data, max_bins, subsample, random_state):
                                       interpolation='midpoint').astype(X_DTYPE)
             assert midpoints.shape[0] == max_bins - 1
 
+        # We avoid having +inf thresholds: +inf thresholds are only allowed in
+        # a "split on nan" situation.
+        np.clip(midpoints, a_min=None, a_max=1e300, out=midpoints)
```

Comment:
See https://github.com/microsoft/LightGBM/blob/master/include/LightGBM/utils/common.h#L665

Not sure where to properly declare this constant? In `types.pyx`? Renaming it into `common.pyx`?

Reply:
oh, this is kinda dirty :D

Reply:
what does that mean?

Reply:
this basically means if the data has a value larger than this, it'll be treated as `inf`, kinda, isn't it?

From one perspective, instead of having `inf`, `nan`, and a real value as threshold, we have a real value, a magic constant, and `inf`. I'm not sure if the latter is a cleaner solution from the perspective of only looking at the threshold.

Reply:
> this basically means if the data has a value larger than this, it'll be treated as inf, kinda, isn't it?

Yes. Though I doubt we need to worry about the [1e300, +inf] range.

> I'm not sure if the latter is a cleaner solution from the perspective of only looking at the threshold.

It makes the code simpler, from what I can tell. And quite likely faster, since it involves less checks (I haven't checked though).

Note that this part is actually independent from the missing values support this is purely infinite values support related.

Reply:
I ran a quick benchmark and did not notice any striking difference. I'm not sure I can trust it though, results are highly variables on my machine.

I prefer this solution better because:

- It avoids having to check for various edge cases in the code, both in the binner and in the predictors. The code is simpler and easier to read.
- LightGBM does it. They might have other reasons that we ignore. And we've been religiously following their tricks so far.
- One can easily argue that it makes sense to forbid thresholds to be +inf: since the thresholds are the (strict) upper bounds of the bins, having a threshold at +inf would mean that no sample can be mapped to the last bin.

Maybe @ogrisel would have some thoughts, but I think he's away already?

Reply:
Sure, I'm convinced. Looks pretty good now (except that one could never have too much comment in such a codebase).

Seems like we're almost done with this then, aren't we?

Reply:
I think so too. I'll make a last pass on the code and try to reduce the diff if I can.

We still need an example, maybe in another PR?

Reply:
@adrinjalali 	regarding where to hardcode the constant value, would you be OK with renaming `types.pyx` into `common.pyx` and have it here?

Reply:
yeah, makes sense.